lib/evmone/eof.cpp

```diff
@@ -129,45 +422,114 @@ std::variant<EOF1Header, EOFValidationError> validate_eof1(
+            return msh_validation_result.first;
+        if (msh_validation_result.second != header.types[code_idx].max_stack_height)
+            return EOFValidationError::invalid_max_stack_height;
+    }
+
+    return header;
 }
 
+}  // namespace
+
 bool is_eof_container(bytes_view container) noexcept
 {
     return container.size() > 1 && container[0] == MAGIC[0] && container[1] == MAGIC[1];
 }
 
-EOF1Header read_valid_eof1_header(bytes_view container) noexcept
+/// This function expects the prefix and version to be valid, as it ignores it.
+EOF1Header read_valid_eof1_header(bytes_view container)
 {
+    EOFSectionHeaders section_headers;
+    auto it = container.begin() + std::size(MAGIC) + 1;  // MAGIC + VERSION
+    while (*it != TERMINATOR)
+    {
+        const auto section_id = *it++;
+        if (section_id == CODE_SECTION)
+        {
+            const auto code_section_num = read_uint16_be(it);
+            it += 2;
+            for (uint16_t i = 0; i < code_section_num; ++i)
+            {
+                const auto section_size = read_uint16_be(it);
+                it += 2;
+                section_headers[section_id].emplace_back(section_size);
+            }
+        }
+        else
+        {
+            const auto section_size = read_uint16_be(it);
+            it += 2;
+            section_headers[section_id].emplace_back(section_size);
+        }
+    }
+    const auto header_size = eof_header_size(section_headers);
+
     EOF1Header header;
-    const auto code_size_offset = 4;  // MAGIC + VERSION + CODE_SECTION_ID
-    header.code_size =
-        static_cast<uint16_t>((container[code_size_offset] << 8) | container[code_size_offset + 1]);
-    if (container[code_size_offset + 2] == 2)  // is data section present
+
+    if (section_headers[TYPE_SECTION].empty())
+        header.types.emplace_back(0, 0, 0);
+    else
     {
-        const auto data_size_offset = code_size_offset + 3;
-        header.data_size = static_cast<uint16_t>(
-            (container[data_size_offset] << 8) | container[data_size_offset + 1]);
+        for (auto type_offset = header_size;
+             type_offset < header_size + section_headers[TYPE_SECTION][0]; type_offset += 4)
+
+            header.types.emplace_back(container[type_offset], container[type_offset + 1],
+                read_uint16_be(&container[type_offset + 2]));
     }
+
+    header.code_sizes = section_headers[CODE_SECTION];
+    std::vector<uint16_t> code_offsets;
+    auto code_offset =
+        header_size +
+        (section_headers[TYPE_SECTION].empty() ? uint16_t{0} : section_headers[TYPE_SECTION][0]);
+    for (const auto code_size : header.code_sizes)
+    {
+        assert(code_offset <= std::numeric_limits<uint16_t>::max());
+        header.code_offsets.emplace_back(static_cast<uint16_t>(code_offset));
+        code_offset += code_size;
+    }
+
+    header.data_size =
+        section_headers[DATA_SECTION].empty() ? uint16_t{0} : section_headers[DATA_SECTION][0];
```

Comment:
Data section is mandatory
```suggestion
        section_headers[DATA_SECTION][0];
```